### PR TITLE
MDRset & MDRpc with gameclock. Fixed network master setting

### LINKED ISSUE #32: Feature: Replication Group Manager
Finish the replication group manager, this is a followup from #17 

  - Need to put interval updated properties into separate update groups at different frames
  - Should load balance automatically

### LINKED ISSUE #33: Feature: Custom MDRpc calls that use game clock
Instead of OneShot mode for [MDReplicated] attribute instead we will add a custom MDRpc that can use the game clock and ensure the message arrives at the correct time. This is a followup to #17

- Custom MDRpc calls that take game clock into account
- This is the solution to OneShot replication

### LINKED ISSUE #38: Allow for swapping of network master
After my refactoring network master can't be swapped since we buffer this information instead of checking every update.

```csharp
public virtual bool ShouldReplicate()
{
    // Note this means we don't support nodes swapping network master.
    // If that should be supported then CheckIfShouldReplicate has to be run every update.
    return IsShouldReplicate;
}
```

It would be good to provide a solution that supports buffering but also allows for switching of network master. Perhaps we can add a node extension along the same line of **SpawnNetworkNode()** that is called something like **ChangeNetworkMaster(int PeerId)** which would simply change the network master of any node you call it on.

### PR DESCRIPTION
**Pull request by Beider**
This fixes #33 and #38 
I don't know why all my previous commits keep showing up in the list here, I even pulled back from master to my branch but it still shows all the old commits.

Anyway I had to change the synchronizer a bit since it was using MDRpcServer which cause the game synchronizer to use the game clock during synchronization, which in turn caused it to crash.

### CODE REVIEW

**Comment by Beider** [comment]
I forgot to add something for the Rset that I had to add in a second commit. Updated and should be complete now.

**Comment by Beider** [comment]
I added the replication group manager as well now, I did not fix a max cap for replication, not sure how to do this in a good way. 

But this should now also fix #32

**Comment by DoubleDeez** [comment]
> I don't know why all my previous commits keep showing up in the list here, I even pulled back from master to my branch but it still shows all the old commits.

I think you need to force push this branch into yours (or clone a fresh repo) and then use feature branches instead of working in master to prevent that.

**Comment by Beider** [comment]
I guess I will try that once these changes are merged. Usually too lazy to do feature branches when I am the only commiting on my fork, but I guess if it helps avoid this whole mess I should do it.

I will fix any outstanding issues you got with this request tomorrow, getting late here. I want to start working on the network list replication next as that is the last feature I need to keep working on my own game.

File: MDExtensions/MDNodeExtensions.cs [hunk 1]
```diff
@@ -298,6 +347,31 @@ public static object MDServerRpcUnreliable(this Node Instance, string Method, pa
         return Instance.MDRpcUnreliableId(ServerId, Method, Args);
     }
 
+    public static bool Invoke(this Node Instance, String Method, params object[] Parameters)
+    {
+        Type NodeType = Instance.GetType();
+        MethodInfo Info = NodeType.GetMethod(Method, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
```
Comment (DoubleDeez) [inline MDExtensions/MDNodeExtensions.cs:353]: I think this is where FastMember would help as this is a lookup by string. Something we can look into in the future.

File: MDNetworking/Misc/MDReplicatorGroupManager.cs [hunk 2]
```diff
@@ -4,9 +4,94 @@
...
+        }
+
+        foreach (HashSet<MDReplicatedMember> group in ReplicationGroups)
+        {
+            if (group.Contains(Member))
+            {
+                group.Remove(Member);
+                break;
+            }
+        }
```
Comment (DoubleDeez) [inline MDNetworking/Misc/MDReplicatorGroupManager.cs:68]: Might be worth adding a "Rebalance" check here eventually.

**Review by DoubleDeez — APPROVED** [review]

**Pull request merged**